### PR TITLE
Normalize late-bound projections with inference variables in fulfillment

### LINKED ISSUE #90729: Problem with GAT and HRTB
I tried this code:

```rust
#![feature(generic_associated_types)]

use std::marker::PhantomData;

pub trait Type {
    type Ref<'a>;
}

pub trait AsBytes {
    fn as_bytes(&self) -> &[u8];
}

impl AsBytes for &str {
    fn as_bytes(&self) -> &[u8] {
        (*self).as_bytes()
    }
}

pub struct Utf8;

impl Type for Utf8 {
    type Ref<'a> = &'a str;
}

pub struct Bytes<T: Type> {
    _marker: PhantomData<T>,
}

impl<T: Type> Bytes<T>
where
    for<'a> T::Ref<'a>: AsBytes,
{
    pub fn new() -> Self {
        Self {
            _marker: PhantomData,
        }
    }
}

fn main() {
    let _b = Bytes::<Utf8>::new();
}
```

I expected to see this happen: *I had expected compilation to succeed*

Instead, this happened: *Compilation fails with the error message below.*

```
error[E0277]: the trait bound `for<'a> <Utf8 as Type>::Ref<'a>: AsBytes` is not satisfied
  --> src\main.rs:41:14
   |
41 |     let _b = Bytes::<Utf8>::new();
   |              ^^^^^^^^^^^^^^^^^^ the trait `for<'a> AsBytes` is not implemented for `<Utf8 as Type>::Ref<'a>`
   |
note: required by `Bytes::<T>::new`
  --> src\main.rs:33:5
   |
33 |     pub fn new() -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^
help: consider further restricting the associated type
   |
40 | fn main() where for<'a> <Utf8 as Type>::Ref<'a>: AsBytes {
   |           ++++++++++++++++++++++++++++++++++++++++++++++
```

### Meta

`rustc --version --verbose`:
```
rustc 1.58.0-nightly (e269e6bf4 2021-10-26)
binary: rustc
commit-hash: e269e6bf47f40c9046cd44ab787881d700099252
commit-date: 2021-10-26
host: x86_64-pc-windows-msvc
release: 1.58.0-nightly
LLVM version: 13.0.0
```

### LINKED ISSUE #93341: Failed to unify types when using GATs and HRTBs
I believe the following code should typecheck under nightly (https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f3d45f199cadea62341a727f66c52d86):

```rust
#![feature(generic_associated_types)]
use std::marker::PhantomData;

struct Id<'id>(PhantomData<fn(&'id ()) -> &'id ()>);

fn new_id() -> Id<'static> {
    Id(PhantomData)
}

pub trait HasLifetime where {
    type AtLifetime<'a>;
}

pub struct ExistentialLifetime<S: HasLifetime>(S::AtLifetime<'static>);

impl<S: HasLifetime> ExistentialLifetime<S> {
    pub fn new<F>(f: F) -> ExistentialLifetime<S>
        where for<'id> F: FnOnce(Id<'id>) -> S::AtLifetime<'id> {
        ExistentialLifetime(f(new_id()))
    }
}


struct ExampleS<'id>(Id<'id>);

struct ExampleMarker;

impl HasLifetime for ExampleMarker {
    type AtLifetime<'id> = ExampleS<'id>;
}


fn broken0() -> ExistentialLifetime<ExampleMarker> {
    fn new_helper<'id>(id: Id<'id>) -> ExampleS<'id> {
        ExampleS(id)
    }

    ExistentialLifetime::<ExampleMarker>::new(new_helper)
}

fn broken1() -> ExistentialLifetime<ExampleMarker> {
    fn new_helper<'id>(id: Id<'id>) -> <ExampleMarker as HasLifetime>::AtLifetime<'id> {
        ExampleS(id)
    }

    ExistentialLifetime::<ExampleMarker>::new(new_helper)
}

fn broken2() -> ExistentialLifetime<ExampleMarker> {
    ExistentialLifetime::<ExampleMarker>::new(|id| ExampleS(id))
}
```

Note that this uses generic associated types (`AtLifetime<'a>`) and it also uses a higher-rank trait bound in `fn new`.

However, rustc nightly rejects this program with the following error:

```
 Compiling playground v0.0.1 (/playground)
error[E0271]: type mismatch resolving `for<'id> <for<'id> fn(Id<'id>) -> ExampleS<'id> {broken0::new_helper} as FnOnce<(Id<'id>,)>>::Output == <ExampleMarker as HasLifetime>::AtLifetime<'id>`
  --> src/lib.rs:38:5
   |
38 |     ExistentialLifetime::<ExampleMarker>::new(new_helper)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected associated type, found struct `ExampleS`
   |
   = note: expected associated type `<ExampleMarker as HasLifetime>::AtLifetime<'_>`
                       found struct `ExampleS<'_>`
   = help: consider constraining the associated type `<ExampleMarker as HasLifetime>::AtLifetime<'_>` to `ExampleS<'_>` or calling a method that returns `<ExampleMarker as HasLifetime>::AtLifetime<'_>`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
note: required by a bound in `ExistentialLifetime::<S>::new`
  --> src/lib.rs:18:46
   |
17 |     pub fn new<F>(f: F) -> ExistentialLifetime<S>
   |            --- required by a bound in this
18 |         where for<'id> F: FnOnce(Id<'id>) -> S::AtLifetime<'id> {
   |                                              ^^^^^^^^^^^^^^^^^^ required by this bound in `ExistentialLifetime::<S>::new`

error[E0271]: type mismatch resolving `for<'id> <for<'id> fn(Id<'id>) -> <ExampleMarker as HasLifetime>::AtLifetime<'id> {broken1::new_helper} as FnOnce<(Id<'id>,)>>::Output == <ExampleMarker as HasLifetime>::AtLifetime<'id>`
  --> src/lib.rs:46:5
   |
46 |     ExistentialLifetime::<ExampleMarker>::new(new_helper)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected associated type, found struct `ExampleS`
   |
   = note: expected associated type `<ExampleMarker as HasLifetime>::AtLifetime<'_>`
                       found struct `ExampleS<'_>`
   = help: consider constraining the associated type `<ExampleMarker as HasLifetime>::AtLifetime<'_>` to `ExampleS<'_>` or calling a method that returns `<ExampleMarker as HasLifetime>::AtLifetime<'_>`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
note: required by a bound in `ExistentialLifetime::<S>::new`
  --> src/lib.rs:18:46
   |
17 |     pub fn new<F>(f: F) -> ExistentialLifetime<S>
   |            --- required by a bound in this
18 |         where for<'id> F: FnOnce(Id<'id>) -> S::AtLifetime<'id> {
   |                                              ^^^^^^^^^^^^^^^^^^ required by this bound in `ExistentialLifetime::<S>::new`

error[E0271]: type mismatch resolving `for<'id> <[closure@src/lib.rs:50:47: 50:64] as FnOnce<(Id<'id>,)>>::Output == <ExampleMarker as HasLifetime>::AtLifetime<'id>`
  --> src/lib.rs:50:5
   |
50 |     ExistentialLifetime::<ExampleMarker>::new(|id| ExampleS(id))
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected associated type, found struct `ExampleS`
   |
   = note: expected associated type `<ExampleMarker as HasLifetime>::AtLifetime<'_>`
                       found struct `ExampleS<'_>`
   = help: consider constraining the associated type `<ExampleMarker as HasLifetime>::AtLifetime<'_>` to `ExampleS<'_>` or calling a method that returns `<ExampleMarker as HasLifetime>::AtLifetime<'_>`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
note: required by a bound in `ExistentialLifetime::<S>::new`
  --> src/lib.rs:18:46
   |
17 |     pub fn new<F>(f: F) -> ExistentialLifetime<S>
   |            --- required by a bound in this
18 |         where for<'id> F: FnOnce(Id<'id>) -> S::AtLifetime<'id> {
   |                                              ^^^^^^^^^^^^^^^^^^ required by this bound in `ExistentialLifetime::<S>::new`

For more information about this error, try `rustc --explain E0271`.
error: could not compile `playground` due to 3 previous errors
```

I think rustc is wrong to reject this code. As you can see, it fails to unify type `<ExampleMarker as HasLifetime>::AtLifetime<'_>` with type `ExampleS<'_>`. But these types _should_ unify!

Version of rustc: `1.60.0-nightly 2022-01-25 8cdb3cd94efece1e17cb`.

### LINKED ISSUE #93342: Confusing output for signature mismatch in Fn trait with GAT lifetime parameter
I tried this code:

```rust
#![feature(generic_associated_types)]

use std::marker::PhantomData;

pub trait Scalar: 'static {
    type RefType<'a>: ScalarRef<'a>;
}

pub trait ScalarRef<'a>: 'a {}

impl Scalar for i32 {
    type RefType<'a> = i32;
}

impl Scalar for String {
    type RefType<'a> = &'a str;
}

impl Scalar for bool {
    type RefType<'a> = i32;
}

impl<'a> ScalarRef<'a> for bool {}

impl<'a> ScalarRef<'a> for i32 {}

impl<'a> ScalarRef<'a> for &'a str {}

fn str_contains(a: &str, b: &str) -> bool {
    a.contains(b)
}

pub struct BinaryExpression<A: Scalar, B: Scalar, O: Scalar, F>
where
    F: Fn(A::RefType<'_>, B::RefType<'_>) -> O,
{
    f: F,
    _phantom: PhantomData<(A, B, O)>,
}

impl<A: Scalar, B: Scalar, O: Scalar, F> BinaryExpression<A, B, O, F>
where
    F: Fn(A::RefType<'_>, B::RefType<'_>) -> O,
{
    pub fn new(f: F) -> Self {
        Self {
            f,
            _phantom: PhantomData,
        }
    }
}

fn fail_case() {
    //     error[E0631]: type mismatch in function arguments
    //   --> src/main.rs:58:54
    //    |
    // 33 | fn str_contains(a: &str, b: &str) -> bool {
    //    | ----------------------------------------- found signature of `for<'r, 's> fn(&'r str, &'s str) -> _`
    // ...
    // 58 |     BinaryExpression::<String, String, bool, _>::new(str_contains);
    //    |     ------------------------------------------------ ^^^^^^^^^^^^ expected signature of `for<'r, 's> fn(<String as Scalar>::RefType<'r>, <String as Scalar>::RefType<'s>) -> _`
    //    |     |
    //    |     required by a bound introduced by this call
    //    |
    // note: required by a bound in `BinaryExpression::<A, B, O, F>::new`
    //   --> src/main.rs:47:8
    //    |
    // 47 |     F: Fn(A::RefType<'_>, B::RefType<'_>) -> O,
    //    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `BinaryExpression::<A, B, O, F>::new`
    // 48 | {
    // 49 |     pub fn new(f: F) -> Self {
    //    |            --- required by a bound in this

    BinaryExpression::<String, String, bool, _>::new(str_contains);
}

fn success_case() {
    BinaryExpression::<String, String, bool, _> {
        f: str_contains,
        _phantom: PhantomData,
    };
}

fn main() {
    println!("Hello, world!");
}
```

I expected to see this happen:

Both functions compile.

Instead, this happened:

If I directly call `new`,

```
expected signature of `for<'r, 's> fn(<String as Scalar>::RefType<'r>, <String as Scalar>::RefType<'s>) -> _`
```

... but if I manually construct the struct, it succeeds.

### Meta

`rustc --version --verbose`:
```
rustc 1.60.0-nightly (8cdb3cd94 2022-01-25)
binary: rustc
commit-hash: 8cdb3cd94efece1e17cbd8f6edb1dc1a482779a0
commit-date: 2022-01-25
host: aarch64-apple-darwin
release: 1.60.0-nightly
LLVM version: 13.0.0
```

Thanks for investigating!

### LINKED ISSUE #94160: Associated type in higher-rank trait bounds cause unexpected E0227 and useless help
Given the following code: [code in playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=edb7c5bf275b057585f888024d52aaa5) 

```rust
use std::ops::Add;

fn add<T>(a: T, b: T) -> T
where
    T: for<'x> From<<&'x T as Add>::Output>,
    for<'x> &'x T: Add,
{
    T::from(&a + &b)
}

fn main() {
    add(1i32, 2i32); // E0227 with useless help
    //add::<i32>(1i32, 2i32); // works
}
```

The current output is:

```
error[E0277]: the trait bound `for<'x> i32: From<<&'x i32 as Add>::Output>` is not satisfied
  --> src/main.rs:12:5
   |
12 |     add(1i32, 2i32); // E0227 with useless help
   |     ^^^ the trait `for<'x> From<<&'x i32 as Add>::Output>` is not implemented for `i32`
   |
note: required by a bound in `add`
  --> src/main.rs:5:8
   |
3  | fn add<T>(a: T, b: T) -> T
   |    --- required by a bound in this
4  | where
5  |     T: for<'x> From<<&'x T as Add>::Output>,
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `add`
help: consider introducing a `where` bound, but there might be an alternative better way to express this requirement
   |
11 | fn main() where i32: for<'x> From<<&'x i32 as Add>::Output> {
   |           +++++++++++++++++++++++++++++++++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground` due to previous error
```

Ideally the help should look like:

```
help: consider introducing the explicit type `::<i32>`
   |
12 |     add::<i32>(1i32, 2i32); // E0227 with useless help
   |        +++++++
```

Not *higher-rank* trait bounds (below code) is compiled successfully.
```rust
use std::ops::Add;

fn add<T>(a: T, b: T) -> T
where
    T: From<<T as Add>::Output> + Add,
{
    T::from(a + b)
}

fn main() {
    add(1i32, 2i32); // works
}
```

### PR DESCRIPTION
**Pull request by compiler-errors**
I modified the PR #90887 <i>"Try to normalize associated types before processing obligations"</i> to only attempt normalization when these two conditions are both true:
1. the projection type has escaping late-bound variables, and 
2. the projection has remaining inference variables.

The only projections that we expect to see inside of a predicate during fulfillment (and which we have any hope of normalizing further) are those that have remaining inference variables that need to be resolved, and which have late-bound vars which opted them out of being replaced with a fresh inference variable in the first place. We rely on existing normalization that is speckled through typeck and confirmation (e.g. [here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_trait_selection/src/traits/select/mod.rs#L2339)) to uphold this invariant, which means we can skip a large portion of the normalization calls that we saw in #90887.

As seen in the (really messy) initial PR I put up for a perf run, it [seems to have improved performance](https://github.com/rust-lang/rust/pull/94042#issuecomment-1041303420).

----

Reasoning about why these two conditions together are sufficient: 
1. If we see a projection during the fulfillment loop that has inference but no late-bound variables, we probably are missing a normalization call somewhere else, since that projection should've been replaced with an inference variable when we got the type from HIR. This means we should track down that missing normalization call to somewhere in typeck or confirmation.
2. If we see a projection with escaping late-bound variables (but no inference variables), then we probably don't have a solution to that projection variable in the first place, since we should have been able to normalize it (since we have all the information we can have about the type).

This probably needs more discussion, but I'm putting this up for initial review, and we can perhaps discuss it in the GATs meeting.

r? @nikomatsakis 
cc @jackh726 

Fixes #90729
Fixes #93341
Fixes #93342
Fixes #94160

### CODE REVIEW

**Comment by jackh726** [comment]
Things to think about:
* I liked removing the extra normalizations in confirmations; it really highlights that this actually simplifies things (this also in theory could have been the reason for better perf, we'll see)
* Was moving the normalization in fulfill up a bit enough to fix perf? In other words, do we *really* need to only normalize if there are late-bound vars?

**Comment by compiler-errors** [comment]
Yeah, regard the first part, I actually discovered that all that extra normalizing of those Fn trait refs wasn't needed when I stopped first shallow resolving and then normalizing during fulfillment... hence this PR is a lot shorter.

The second question is a good point, I can put up a quick PR tomorrow unless you want to modify #90887 yourself.

**Comment by compiler-errors** [comment]
Haha! That's lovely. Let me rebase that normalization commit when I get up in a few hours.

**Comment by compiler-errors** [comment]
Interesting to see that the perf improvement is a combination of the normalization in selection _and_ (probably?) the type flag addition.

Per discussion with @jackh726, let's test this hypothesis further. I reverted the type flag diff (and added a small change so we don't resolve variables _and_ normalize in fulfillment, since that's kinda doing duplicated work).

File: compiler/rustc_trait_selection/src/traits/fulfill.rs [hunk 1]
```diff
@@ -344,17 +344,32 @@ impl<'a, 'b, 'tcx> FulfillProcessor<'a, 'b, 'tcx> {
...
+                obligation.cause.clone(),
+                obligation.recursion_depth + 1,
+                obligation.predicate,
+                &mut obligations,
+            );
+            if predicate != obligation.predicate {
+                obligations.push(obligation.with(predicate));
+                return ProcessResult::Changed(mk_pending(obligations));
+            }
+        } else if obligation.predicate.has_infer_types_or_consts() {
```
Comment (jackh726) [inline compiler/rustc_trait_selection/src/traits/fulfill.rs:366]: Just fyi this now skips the inference var resolution if there is a projection but we can't normalize.

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/fulfill.rs:366]: I was thinking about that just now... Do you want me to revert and we can re-queue the test, or should we just do perf on whatever's currently queued?

**Comment by nikomatsakis** [comment]
@compiler-errors do you plan to rebase atop https://github.com/rust-lang/rust/pull/90887 ?

**Comment by jackh726** [comment]
Not sure the slight perf improvement here is worth the added complexity.

**Comment by compiler-errors** [comment]
Yeah, in that case, I'll close this out.

**Pull request closed without merging**